### PR TITLE
Implement NSProxy

### LINKED ISSUE #2348: Add NSProxy support
All of this class is entirely stubbed out, its time we had them for real.

### PR DESCRIPTION
**Pull request by ms-jihua**
Fixes #2348

### CODE REVIEW

File: Frameworks/Foundation/NSProxy.mm [hunk 1]
```diff
@@ -14,220 +14,317 @@
...
 }
 
 /**
- @Status Stub
+ @Status Interoperable
  @Notes
 */
 - (void)forwardInvocation:(NSInvocation*)anInvocation {
-    UNIMPLEMENTED();
+    [NSException raise:NSInvalidArgumentException
```
Comment (aballway) [inline Frameworks/Foundation/NSProxy.mm]: nit: should these just be NSInvalidAbstractInvocationReturn() ? I know it's not a class cluster per se, but it's still "Ya gotta override this"

File: Frameworks/Foundation/NSProxy.mm [hunk 2]
```diff
@@ -14,220 +14,317 @@
...
- @Status Stub
+ @Status Interoperable
  @Notes
 */
 - (oneway void)release {
-    UNIMPLEMENTED();
+    intptr_t* refCount = ((intptr_t*)self) - 1;
+    // We allow refcounts to run into the negative, but should only
+    // deallocate once.
+    if (__sync_sub_and_fetch(refCount, 1) == -1) {
```
Comment (aballway) [inline Frameworks/Foundation/NSProxy.mm]: should this be <= ? Ideally we'd never go below -1 but in case we ever should due to async nonsense we wouldn't want to leak

File: Frameworks/Foundation/NSProxy.mm (hunk 1 above)
Comment (ms-jihua) [inline Frameworks/Foundation/NSProxy.mm]: nope, documented NSInvalidArgumentException: https://developer.apple.com/reference/foundation/nsproxy/1416417-forwardinvocation?language=objc

File: Frameworks/Foundation/NSProxy.mm (hunk 2 above)
Comment (ms-jihua) [inline Frameworks/Foundation/NSProxy.mm]: - this is just lifted from NSObject.mm
- async nonsense wouldn't be able to cause that since __sync_sub_and_fetch is atomic

File: Frameworks/Foundation/NSObject.mm [hunk 3]
```diff
@@ -708,7 +707,8 @@ void WinObjC_SetMissingSelectorFatal(BOOL fatal) {
 }
 @end
 
-static void _dealloc_zombify(id self, SEL _cmd); // forward declaration
+    static void
```
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSObject.mm]: nit: I'd prefer not having the format changes in this PR

File: Frameworks/Foundation/NSObject.mm [hunk 4]
```diff
@@ -737,7 +737,7 @@ + (BOOL)isSubclassOfClass:(Class)classRef {
 /**
  @Status Interoperable
 */
-- (void)doesNotRecognizeSelector: (SEL)selector {
+- (void)doesNotRecognizeSelector:(SEL)selector {
```
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSObject.mm]: then again, some of them are good ..

File: Frameworks/Foundation/NSProxy.mm [hunk 5]
```diff
@@ -14,220 +14,317 @@
...
 
 /**
- @Status Stub
+ @Status Interoperable
  @Notes
 */
 - (NSMethodSignature*)methodSignatureForSelector:(SEL)aSelector {
-    UNIMPLEMENTED();
-    return StubReturn();
+    [NSException raise:NSInvalidArgumentException
```
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSProxy.mm]: nit: macro these, and use `%hs` + `__PRETTY_FUNCTION__` to get the method name in the macro

File: Frameworks/Foundation/NSProxy.mm [hunk 6]
```diff
@@ -14,220 +14,317 @@
...
-     @Notes
-    */
-    - (Class) class {
-    UNIMPLEMENTED();
-    return StubReturn();
+/**
+ @Status Interoperable
+ @Notes
+*/
+- (Class)superclass {
```
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSProxy.mm]: Not sure on whether this should be calling `[self class]`; thoughts?

File: Frameworks/Foundation/NSProxy.mm [hunk 7]
```diff
@@ -14,220 +14,317 @@
...
+}
+
+/**
+ @Status Interoperable
  @Notes
 */
 - (instancetype)self {
-    UNIMPLEMENTED();
-    return StubReturn();
+    return (id)self;
```
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSProxy.mm]: nit: cast necessary?

File: Frameworks/Foundation/NSProxy.mm [hunk 8]
```diff
@@ -14,220 +14,317 @@
...
 
 /**
- @Status Stub
+ @Status Interoperable
  @Notes
 */
 - (BOOL)isKindOfClass:(Class)aClass {
-    UNIMPLEMENTED();
-    return StubReturn();
+    Class forwardClass = static_cast<Class>([self performSelector:@selector(class)]);
```
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSProxy.mm]: aaa is it really guaranteed that it calls `forwardInvocation:` for `class`?

Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSProxy.mm]: i'm not sure I like that this is different from calling `class`

File: Frameworks/Foundation/NSProxy.mm [hunk 9]
```diff
@@ -14,220 +14,317 @@
...
 
 /**
- @Status Stub
+ @Status Interoperable
  @Notes
 */
 - (id)performSelector:(SEL)aSelector withObject:(id)anObject {
-    UNIMPLEMENTED();
-    return StubReturn();
+    NSInvocation* invocation = [NSInvocation invocationWithMethodSignature:[self methodSignatureForSelector:aSelector]];
```
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSProxy.mm]: it's strange to me that these go through invocation instead of hitting the invocation-forwarding path

File: Frameworks/Foundation/NSProxy.mm [hunk 10]
```diff
@@ -14,220 +14,317 @@
...
+ *  opt out of the fast path, and must implement both retain and release.
+ *
+ *  To actually be ARC compliant, we need to store our refcount sizeof(intptr_t) bytes before
+ *  self->isa, and mimic the runtime's behaviour for direct invocations of retain/release.
+ */
+- (void)_ARCCompliantRetainRelease {
 }
 
-@end
+@end
```
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSProxy.mm]: you may have to implement doesNotRecognizeSelector:, and call it from the default forwardInvocation: implementation just like NSObject -- that's how people will get missing selector exceptions

File: include/Foundation/NSProxy.h [hunk 11]
```diff
@@ -22,19 +22,25 @@
...
 + (id)alloc;
 + (id)allocWithZone:(NSZone*)zone;
 - (void)dealloc;
 - (void)finalize;
 - (void)forwardInvocation:(NSInvocation*)anInvocation;
 - (NSMethodSignature*)methodSignatureForSelector:(SEL)aSelector;
 + (BOOL)respondsToSelector:(SEL)aSelector;
-+ (Class) class;
++ (Class)class;
++ (Class)superclass; // Not in doc, but is supported
```
Comment (DHowett-MSFT) [inline include/Foundation/NSProxy.h]: Doesn't the NSObject protocol contain this?

File: tests/unittests/Foundation/NSProxyTests.mm [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+- (NSUInteger)count;
+- (id)objectAtIndex:(NSUInteger)index;
+- (void)addObject:(id)anObject;
+- (void)insertObject:(id)anObject atIndex:(NSUInteger)index;
+@end
+
+@implementation _NSTestMutableArrayProxy
+
+- (instancetype)initWithMutableArray:(NSMutableArray*)array {
+    self->_forwardArray = array;
```
Comment (DHowett-MSFT) [inline tests/unittests/Foundation/NSProxyTests.mm]: nit: document not calling super init

File: Frameworks/Foundation/NSProxy.mm (hunk 8 above)
Comment (ms-jihua) [inline Frameworks/Foundation/NSProxy.mm]: it's the contract unfortunately :( [NSProxy class] does not return a forwarded class

File: include/Foundation/NSProxy.h (hunk 11 above)
Comment (ms-jihua) [inline include/Foundation/NSProxy.h]: it does not!

File: tests/unittests/Foundation/NSProxyTests.mm (hunk 12 above)
Comment (ms-jihua) [inline tests/unittests/Foundation/NSProxyTests.mm]: can't/shouldn't call super init on NSProxy - it's NS_UNAVAILABLE (this matches the reference platform)

File: Frameworks/Foundation/NSProxy.mm (hunk 9 above)
Comment (ms-jihua) [inline Frameworks/Foundation/NSProxy.mm]: what do you mean?

Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSProxy.mm]: It seems like `performSelector:` and friends _could_ just call objc_msgSend; that would go through the invocation forwarding and capture codepath

It's less verbose, but perhaps there is an advantage to creating the invocation directly!

File: tests/unittests/Foundation/NSProxyTests.mm (hunk 12 above)
Comment (DHowett-MSFT) [inline tests/unittests/Foundation/NSProxyTests.mm]: Indeed! It'd be good to have a comment in here so somebody doesn't think it's broken.
Our CC (or that one really didactic talk I gave) states that we should _almost never_ not call init. Since this is one of the cases where we mustn't, it deserves callout!

File: Frameworks/Foundation/NSObject.mm (hunk 3 above)
Comment (ms-jihua) [inline Frameworks/Foundation/NSObject.mm]: i'll fix this one and leave the rest, if that's ok

File: Frameworks/Foundation/NSProxy.mm (hunk 6 above)
Comment (ms-jihua) [inline Frameworks/Foundation/NSProxy.mm]: This is also the same as NSObject.mm... going to propose we leave both for the moment and let sleeping beasts lie.

Comment (ms-jihua) [inline Frameworks/Foundation/NSProxy.mm]: I changed my mind and got curious how the reference platform behaved and it appears superclass is polymorphic so I will make this change in both places,

Comment (aballway) [inline Frameworks/Foundation/NSProxy.mm]: May want to do the fix for NSObject in a separate PR.  Doubt it would cause any regressions but it's foundational enough that it might and should be thoroughly tested.

File: Frameworks/Foundation/NSProxy.mm (hunk 10 above)
Comment (ms-jihua) [inline Frameworks/Foundation/NSProxy.mm]: This is already tested for in the unit test - I believe it forwards to the remote doesNotRecognizeSelector.

File: Frameworks/Foundation/NSProxy.mm (hunk 9 above)
Comment (ms-jihua) [inline Frameworks/Foundation/NSProxy.mm]: discussed offline, leaving as is

File: Frameworks/Foundation/NSProxy.mm (hunk 7 above)
Comment (ms-jihua) [inline Frameworks/Foundation/NSProxy.mm]: seems not

File: Frameworks/Foundation/NSProxy.mm (hunk 6 above)
Comment (ms-jihua) [inline Frameworks/Foundation/NSProxy.mm]: Mm, my thought is that if a change like that breaks something, we'll either be able to catch it in CI build, or it's so incredibly subtle despite being in such an exercised class that doing it in a separate change won't really gain us anything, so I'll elect to risk it. (I will probably regret saying this a week from now)

**Comment by ms-jihua** [comment]
Could not repro CI build failure locally, queuing again.
